javascript/frameworks/ui5/lib/advanced_security/javascript/frameworks/ui5/UI5WebcomponentsReact.qll

```diff
@@ -0,0 +1,58 @@
+import semmle.javascript.security.dataflow.DomBasedXssCustomizations
+
+/**
+ * ``` javascript
+ * import { Input, Button } from '@ui5/webcomponents-react';
+ * ```
+ */
+class WebComponentImport extends DataFlow::SourceNode {
+  string typeName;
+
+  WebComponentImport() {
+    this = API::moduleImport("@ui5/webcomponents-react").getMember(typeName).asSource()
+  }
+
+  string getTypeName() { result = typeName }
+}
+
+/**
+ * Refers to the ref attribute
+ * <SomeElement ref={x}>
+ */
+class RefAttribute extends JsxAttribute {
+  RefAttribute() { this.getName() = "ref" }
+}
+
+/**
+ * Holds if the ref variable is assigned to a UI5 component via JSX
+ */
+predicate isRefAssignedToUI5Component(UseRefDomValueSource source) {
+  exists(
+    Variable refVar, JsxElement jsx, RefAttribute attr, VarRef componentVar, WebComponentImport decl
+  |
+    source.getElement() = jsx and
+    // The JSX element uses a UI5 webcomponent for react
+    jsx.getNameExpr() = componentVar and
+    decl.asExpr() = componentVar.getVariable().getADefinition() and
+    // The JSX element has a ref attribute pointing to our ref variable
+    jsx.getAnAttribute() = attr and
+    attr.getValue().(VarRef).getVariable() = refVar
+  )
+}
```

Comment:
I guess you wanted to express this:

> The UI5 component has a variable assigned to it and its `current` property is read somewhere.

... but there is no easy way to associate `Input` in the `import` statement and the `Input` in the component.

Well, `React::ReactComponent.ref/0` actually does that cleanly; but we can't use it until we make `UI5WebComponent` extend `ReactComponent`. Let's mention this in the `Future Works` and raise an issue so that we can come back to it. For now we can let SSA handle it.

Reply:
yeah the purpose of this is definitely to be able to say "this is a UI5 webcomponent for react type of component usage" - 

I agree this may not be the best way, but I cant currently understand what UI5WebComponent would actually be comprised of (how do we know its from webcomponent-react other than its import, which is basically the same as the above)... 

still ! definitely leaving a future note to see if this can be improved, thanks!